worker/leaseexpiry/worker.go

```diff
@@ -0,0 +1,138 @@
+package leaseexpiry
+
+import (
+	"database/sql"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3"
+	"github.com/juju/worker/v3/catacomb"
+
+	"github.com/juju/juju/database"
+)
+
+// Config encapsulates the configuration options for
+// instantiating a new lease expiry worker.
+type Config struct {
+	Clock  clock.Clock
+	Logger Logger
+	DB     *sql.DB
+}
+
+// Validate checks whether the worker configuration settings are valid.
+func (cfg Config) Validate() error {
+	if cfg.Clock == nil {
+		return errors.NotValidf("nil Clock")
+	}
+	if cfg.Logger == nil {
+		return errors.NotValidf("nil Logger")
+	}
+	if cfg.DB == nil {
+		return errors.NotValidf("nil DB")
+	}
+
+	return nil
+}
+
+type expiryWorker struct {
+	catacomb catacomb.Catacomb
+
+	clock  clock.Clock
+	logger Logger
+	db     *sql.DB
+
+	stmt *sql.Stmt
+}
+
+// NewWorker returns a worker that periodically deletes
+// expired leases from the controller database.
+func NewWorker(cfg Config) (worker.Worker, error) {
+	var err error
+
+	if err = cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &expiryWorker{
+		clock:  cfg.Clock,
+		logger: cfg.Logger,
+		db:     cfg.DB,
+	}
+
+	// Prepare our single DML statement before considering the worker started.
+	// Pinned leases may not be deleted even if expired.
+	q := `
+DELETE FROM lease WHERE uuid in (
+    SELECT l.uuid 
+    FROM   lease l LEFT JOIN lease_pin p ON l.uuid = p.lease_uuid
+    WHERE  p.uuid IS NULL
+    AND    l.expiry < datetime('now')
+)`[1:]
```

Comment:
This feels like the wrong location to have SQL. I'm not sure where we would want to move this or even call this from. For now, whilst we're moving to the new DB I think it's fine to leave this here until we see more emergent patterns